### PR TITLE
Upsert Records

### PR DESCRIPTION
**Pull request by sergiogcx**
Closes https://github.com/cityofaustin/atd-data-tech/issues/2194

Code-only review 🙏

### CODE REVIEW

File: atd-etl/app/process/helpers_import.py [hunk 1]
```diff
@@ -122,10 +145,22 @@ def generate_gql(line, fieldnames, file_type):
...
-                                     fields=fields)
+        # Clean field list
+        template_fields = [field.lower() for field in fieldnames if field.lower() not in filters[0][1]]
+        # Generate Template
+        template = generate_template(
+            name=query_name,
+            function=function_name,
+            fields=fields,
+            fieldnames=template_fields,
+            upsert=(file_type != "crash"),
```
Comment (mddilley) [inline atd-etl/app/process/helpers_import.py]: How are upserts for crash records handled? Or are they right now?

File: atd-etl/app/process/helpers_import.py [hunk 2]
```diff
@@ -13,35 +13,58 @@
...
-    """
+    :param str name: The name of the graphql mutation
+    :param str function: The name of the graphql function
+    :param str fields: The value of the fields in graphql expression
+    :param str[] fieldnames: An array of strings containing the names of the columns
+    :param bool upsert: If true, adds upsert logic; false otherwise.
+    :param str constraint: The name of the constraint on_conflict
+    :return str:
+    """
+    if upsert:
```
Comment (mddilley) [inline atd-etl/app/process/helpers_import.py]: Thanks for opening my eyes to upserts in Hasura. I read the docs a little bit just now, and this is making me understand how you are controlling which fields will update even better. 👍

File: atd-etl/app/process/helpers_import.py [hunk 3]
```diff
@@ -122,10 +145,22 @@ def generate_gql(line, fieldnames, file_type):
         fields = generate_fields_with_filters(line=line,
                                               fieldnames=fieldnames,
                                               filters=filters)
-
-        template = generate_template(name=query_name,
-                                     function=function_name,
-                                     fields=fields)
+        # Clean field list
+        template_fields = [field.lower() for field in fieldnames if field.lower() not in filters[0][1]]
```
Comment (mddilley) [inline atd-etl/app/process/helpers_import.py]: I'm out of the loop so I'm not sure why the fields need to be lowered. Could you help me understand more? Are the headers in the .csv uppercase?

Comment (sergiogcx) [reply atd-etl/app/process/helpers_import.py]: Yes, there are a couple reasons for this.

1. For some reason Hasura does not like camel-case for my fields 🤷‍♂it throws an error.
2. There are configurations settings containing column names in lower case in this script, some comparisons are made against those column names, and python is case sensitive when comparing strings. For example:

`filters[0][1]` contains all the needed to be removed.
`[field.lower() for field in fieldnames if field.lower() not in filters[0][1]]` in plain english: give me a lower-case array of all strings in fieldnames as long as they are not in the removed list.

File: atd-etl/app/process/helpers_import.py (hunk 1 above)
Comment (sergiogcx) [reply atd-etl/app/process/helpers_import.py]: No upserts for crashes at the moment, we are going to assume the crash change needs to be reviewed by a human.

There are plains for having automated upserts if the crash has not been edited by a human, but not within the scope of this issue.

File: atd-etl/app/process/helpers_import.py [hunk 4]
```diff
@@ -13,35 +13,58 @@
...
+              update_columns: [
+                %CONFLICT_FIELDS%
+              ]
+            }
+        """.replace(
+            "%CONFLICT_CONSTRAINT%",
+            constraint
+        ).replace(
+            "%CONFLICT_FIELDS%",
+            ",\n".join([f.lower() for f in fieldnames])
```
[the remark this answers is not in the document]
Comment (sergiogcx) [reply atd-etl/app/process/helpers_import.py]: Thanks Mike! 👍

File: atd-etl/app/process/helpers_import.py (hunk 3 above)
Comment (sergiogcx) [reply atd-etl/app/process/helpers_import.py]: Thank you for your questions, I should make the code easier to read  or at least add some commentary

File: atd-etl/app/process/helpers_import.py (hunk 1 above)
Comment (mddilley) [reply atd-etl/app/process/helpers_import.py]: Awesome, thanks!

File: atd-etl/app/process/helpers_import.py (hunk 3 above)
Comment (mddilley) [reply atd-etl/app/process/helpers_import.py]: Thanks for answering them - still familiarizing myself with the whole picture so this was very helpful!

**Comment by sergiogcx** [comment]
Issues fixed, merging to master 👍
